dnd5e.mjs

```diff
@@ -460,8 +467,73 @@ Hooks.once("i18nInit", () => {
   Object.values(CONFIG.DND5E.activityTypes).forEach(c => c.documentClass.localize());
   Object.values(CONFIG.DND5E.advancementTypes).forEach(c => c.documentClass.localize());
   Localization.localizeDataModel(dataModels.settings.TransformationSetting);
+
+  // Set up spellcasting types. This is done between init and setup to allow for modules to register.
+  const spellcastings = Object.entries(DND5E.spellcasting).sort((a, b) => b.order - a.order);
+  for ( const [i, [key, config]] of spellcastings.entries() ) {
+    _handleDeprecatedSpellcasting(key, config);
+
+    const Cls = SpellcastingModel.TYPES[config.type];
+    if ( Cls ) {
+      config.order = 2 * i * Object.keys(CONFIG.DND5E.spellLevels).length;
+      const model = new Cls(config, { key });
+      Object.assign(DND5E.spellcasting, { [key]: model });
+    } else {
+      delete CONFIG.DND5E.spellcasting[key];
+    }
+  }
 });
 
+/* -------------------------------------------- */
+
+/**
+ * Mutate a spellcasting configuration using outdated or missing properties,
+ * attempting to retrieve a table via a hook call.
+ * @param {string} key      The key assigned to the spellcasting config.
+ * @param {object} config   The spellcasting config.
+ */
+function _handleDeprecatedSpellcasting(key, config) {
+  // Fix progression missing type.
+  if ( !config.type ) {
+    foundry.utils.logCompatibilityWarning(
+      `The '${key}' spellcasting should be registered with a 'type' property.`,
+      { from: "5.0", until: "5.4", once: true }
+    );
+    config.type = config.prepares ? "progression" : "static";
+  } else {
+    return;
+  }
+
+  // Fix progression missing table.
+  if ( (config.type !== "static") && !config.table ) {
+    foundry.utils.logCompatibilityWarning(
+      "A registered spellcasting with a progression must define the 'table' property.",
+      { from: "5.0", until: "5.4", once: true }
+    );
+
+    const actor = new CONFIG.Actor.documentClass({ name: "Deprecation", _id: "dnd5eDeprecation", type: "character" });
+    Hooks.call(`dnd5e.prepare${key.capitalize()}Slots`, actor.system.spells, actor, { [key]: 0 });
```

Comment:
Happy to provide a code example in use in a premium module this was tested with. The instantiated actor is only to grab the table that is passed through - it likely does not need to be an actor, though passing in the "correct" parameters is safer.